File: cwe-tests/javascript/cwe_crypto.js
// ===========================================================================
// REACHABLE TEST — DO NOT USE IN PRODUCTION
// CWE-327 (Broken Crypto), CWE-328 (Weak Hash), CWE-330 (Weak PRNG)
// CWE-326 (Insufficient Key Size)
// ===========================================================================
const express = require('express');
const crypto = require('crypto');
const app = express();
app.use(express.json());

// ── REACHABLE: CWE-328 — Weak Hash (MD5 for passwords) ───────────────────
app.post('/api/auth/hash', (req, res) => {
    const password = req.body.password;
    const hash = crypto.createHash('md5').update(password).digest('hex');
    res.json({ hash });
});

app.post('/api/verify/checksum', (req, res) => {
    const data = req.body.data;
    const sha1 = crypto.createHash('sha1').update(data).digest('hex');
    res.json({ sha1 });
});

// ── REACHABLE: CWE-327 — Broken Cipher (DES, RC4) ───────────────────────
app.post('/api/encrypt/des', (req, res) => {
    const key = Buffer.from('8byteskey!!12345'.slice(0, 8));
    const cipher = crypto.createCipheriv('des-ecb', key, null);
    let encrypted = cipher.update(req.body.data, 'utf8', 'hex');
    encrypted += cipher.final('hex');
    res.json({ ciphertext: encrypted });
});

app.post('/api/encrypt/rc4', (req, res) => {
    const key = Buffer.from('weakrc4key');
    const cipher = crypto.createCipheriv('rc4', key, null);
    let encrypted = cipher.update(req.body.data, 'utf8', 'hex');
    encrypted += cipher.final('hex');
    res.json({ ciphertext: encrypted });
});

// ── REACHABLE: CWE-327 — AES-ECB (pattern leak) ─────────────────────────
app.post('/api/encrypt/aes-ecb', (req, res) => {
    const key = Buffer.from('0123456789abcdef0123456789abcdef');
    const cipher = crypto.createCipheriv('aes-256-ecb', key, null);
    let encrypted = cipher.update(req.body.data, 'utf8', 'hex');
    encrypted += cipher.final('hex');
    res.json({ ciphertext: encrypted });
});

// ── REACHABLE: CWE-330 — Weak PRNG for security ─────────────────────────
app.get('/api/token', (req, res) => {
    // BAD: Math.random is not cryptographically secure
    const token = Math.random().toString(36).substring(2) + Math.random().toString(36).substring(2);
    res.json({ token });
});

app.get('/api/session-id', (req, res) => {
    const sid = Array.from({length: 32}, () => 
        'abcdef0123456789'[Math.floor(Math.random() * 16)]
    ).join('');
    res.json({ session_id: sid });
});

app.get('/api/otp', (req, res) => {
    const otp = Math.floor(100000 + Math.random() * 900000);
    res.json({ otp });
});

// ── REACHABLE: CWE-798 — Hardcoded Crypto Keys ──────────────────────────
const HMAC_KEY = 'super-secret-hmac-key-hardcoded-2024';
const AES_KEY = Buffer.from('ThisIsAHardcodedAESKey!1234567890'.slice(0, 32));

app.post('/api/sign', (req, res) => {
    const sig = crypto.createHmac('sha256', HMAC_KEY).update(req.body.data).digest('hex');
    res.json({ signature: sig });
});

app.post('/api/encrypt/hardcoded', (req, res) => {
    const iv = crypto.randomBytes(16);
    const cipher = crypto.createCipheriv('aes-256-cbc', AES_KEY, iv);
    let encrypted = cipher.update(req.body.data, 'utf8', 'hex');
    encrypted += cipher.final('hex');
    res.json({ iv: iv.toString('hex'), ciphertext: encrypted });
});

// ── UNREACHABLE ──────────────────────────────────────────────────────────
function deadMd5() { return crypto.createHash('md5').update('dead').digest('hex'); }
function deadWeakRandom() { return Math.random() * 999999; }

app.listen(4004);
module.exports = app;

File: cwe-tests/python/cwe_crypto.py
# ============================================================================
# REACHABLE TEST FILE — DO NOT USE IN PRODUCTION
# Triggers: Semgrep weak-hash, insecure-cipher, hardcoded-key, weak-random
# CWE-327 (Broken Crypto), CWE-328 (Weak Hash), CWE-330 (Weak PRNG)
# CWE-798 (Hardcoded Credentials in Crypto)
# ============================================================================
"""
Cryptographic weaknesses: weak hashing, broken ciphers, insecure randomness.
"""
from flask import Flask, request, jsonify
import hashlib
import hmac
import random
import string

app = Flask(__name__)


# ============================================================================
# REACHABLE: CWE-328 — Weak Hash (MD5 for passwords)
# ============================================================================
@app.route('/api/auth/register', methods=['POST'])
def register():
    """MD5 for password hashing — trivially reversible."""
    username = request.json.get('username', '')
    password = request.json.get('password', '')
    # BAD: MD5 is not suitable for password hashing
    password_hash = hashlib.md5(password.encode()).hexdigest()
    return jsonify({'user': username, 'hash': password_hash})


# REACHABLE: CWE-328 — Weak Hash (SHA1 for integrity)
@app.route('/api/files/checksum', methods=['POST'])
def file_checksum():
    """SHA1 for file integrity — collision attacks feasible."""
    data = request.get_data()
    # BAD: SHA1 is collision-prone (SHAttered attack, 2017)
    checksum = hashlib.sha1(data).hexdigest()
    return jsonify({'sha1': checksum})


# ============================================================================
# REACHABLE: CWE-327 — Broken Crypto Algorithm
# ============================================================================
@app.route('/api/encrypt', methods=['POST'])
def encrypt_data():
    """DES encryption — 56-bit key, brute-forceable."""
    from Crypto.Cipher import DES
    plaintext = request.json.get('data', '').encode()
    # BAD: DES is broken — 56-bit key
    key = b'8byteky'  # DES requires 8-byte key
    key = key.ljust(8, b'\0')
    cipher = DES.new(key, DES.MODE_ECB)
    # BAD: ECB mode leaks patterns
    padded = plaintext.ljust((len(plaintext) // 8 + 1) * 8, b'\0')
    encrypted = cipher.encrypt(padded)
    return jsonify({'ciphertext': encrypted.hex()})


# REACHABLE: CWE-327 — AES-ECB (pattern leaking)
@app.route('/api/encrypt/aes', methods=['POST'])
def encrypt_aes_ecb():
    """AES-ECB leaks plaintext patterns."""
    from Crypto.Cipher import AES
    plaintext = request.json.get('data', '').encode()
    # BAD: ECB mode — identical plaintext blocks produce identical ciphertext
    key = b'0123456789abcdef'  # Hardcoded AES key
    cipher = AES.new(key, AES.MODE_ECB)
    padded = plaintext.ljust((len(plaintext) // 16 + 1) * 16, b'\0')
    encrypted = cipher.encrypt(padded)
    return jsonify({'ciphertext': encrypted.hex()})


# ============================================================================
# REACHABLE: CWE-330 — Weak PRNG for security-sensitive operations
# ============================================================================
@app.route('/api/token/generate', methods=['GET'])
def generate_token():
    """random.choice for auth tokens — predictable PRNG."""
    # BAD: random module is not cryptographically secure
    token = ''.join(random.choice(string.ascii_letters + string.digits) for _ in range(32))
    return jsonify({'token': token})


@app.route('/api/otp/generate', methods=['GET'])
def generate_otp():
    """random.randint for OTP — predictable."""
    # BAD: Predictable OTP generation
    otp = random.randint(100000, 999999)
    return jsonify({'otp': otp})


# REACHABLE: CWE-330 — Weak random for session ID
@app.route('/api/session/create', methods=['POST'])
def create_session():
    """Predictable session ID using random module."""
    # BAD: Session IDs must use CSPRNG
    session_id = hex(random.getrandbits(128))[2:]
    return jsonify({'session_id': session_id})


# ============================================================================
# REACHABLE: CWE-798 — Hardcoded Cryptographic Key
# ============================================================================
HMAC_SECRET = b"super-secret-hmac-key-never-rotate"  # Hardcoded HMAC key

@app.route('/api/sign', methods=['POST'])
def sign_data():
    """HMAC with hardcoded key."""
    data = request.json.get('data', '').encode()
    # Uses hardcoded key above
    signature = hmac.new(HMAC_SECRET, data, hashlib.sha256).hexdigest()
    return jsonify({'signature': signature})


# ============================================================================
# UNREACHABLE: Same patterns in dead code
# ============================================================================
def _dead_md5_hash():
    """UNREACHABLE — dead code MD5."""
    return hashlib.md5(b"password123").hexdigest()

def _dead_weak_random():
    """UNREACHABLE — dead code weak random."""
    return random.randint(0, 99999)

def _dead_des_encrypt():
    """UNREACHABLE — dead code DES."""
    from Crypto.Cipher import DES
    cipher = DES.new(b'deadkey!', DES.MODE_ECB)
    return cipher.encrypt(b'deaddata')


if __name__ == '__main__':
    app.run(port=5003)
